src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs

```diff
@@ -13,8 +12,32 @@ public abstract partial class Enum
         [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Enum_GetValuesAndNames")]
         private static partial void GetEnumValuesAndNames(QCallTypeHandle enumType, ObjectHandleOnStack values, ObjectHandleOnStack names, Interop.BOOL getNames);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern object InternalBoxEnum(RuntimeType enumType, long value);
+        private static unsafe object InternalBoxEnum(RuntimeType enumType, long value)
+        {
+            MethodTable* pMethodTable = (MethodTable*)enumType.m_handle;
+            object obj = RuntimeHelpers.AllocateUninitializedObject(pMethodTable);
+            ref byte dataRef = ref obj.GetRawData();
+
+            switch (pMethodTable->GetNumInstanceFieldBytes())
+            {
+                case 1:
+                    dataRef = (byte)value;
+                    break;
+                case 2:
+                    Unsafe.As<byte, ushort>(ref dataRef) = (ushort)value;
+                    break;
+                case 3:
+                    Unsafe.As<byte, uint>(ref dataRef) = (uint)value;
+                    break;
+                default:
+                    Unsafe.As<byte, ulong>(ref dataRef) = (ulong)value;
+                    break;
+            }
+
+            GC.KeepAlive(enumType);
```

Comment:
Can that line be moved after the call to `AllocateUninitializedObject`?

Reply:
Do you mean relying on the fact the allocated object would also keep the same method table alive anyway? We still need that in the call to `GetNumInstanceFieldBytes` below that line. Not sure if moving this up would be worth it in terms of making the code more error prone - would leaving this here have much overhead at all? The GC would just keep the object potentially alive for a couple more instructions, but the call itself would still just go away, right? 🤔

Reply:
Yeah, the object holds the type, and the object is held because it will be returned. In either case it would make almost zero difference, this was mostly a cosmetic change.